File: app/src/main/java/com/example/sasham/testproject/themes/ThemesActivity.java
package com.example.sasham.testproject.themes;

import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;

import com.example.sasham.testproject.BaseActivity;
import com.example.sasham.testproject.BaseDaggerActivity;
import com.example.sasham.testproject.Constants;
import com.example.sasham.testproject.R;
import com.example.sasham.testproject.messages.MessagesActivity;
import com.example.sasham.testproject.model.FavoriteThemeInfoRepositoryImp;
import com.example.sasham.testproject.model.Theme;
import com.example.sasham.testproject.util.NetworkUtil;
import com.example.sasham.testproject.util.PreferencesHelper;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

public class ThemesActivity extends BaseDaggerActivity implements ThemesListingFragment.Callback, BaseActivity.OnConnectionListener {

    public static final String SHOW_MESSAGES_ACTION = "show_messages";

    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @BindView(R.id.themes_container)
    CoordinatorLayout mainContainer;

    private Snackbar snackbar;
    private boolean isConnected = false;
    private Fragment themesfragment;

    @Inject
    FavoriteThemeInfoRepositoryImp infoRepository;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_themes);

        ButterKnife.bind(this);
        setSupportActionBar(toolbar);

        setDefaultTitle();

        //Показываем фрагменты только если доступна сеть
        if (NetworkUtil.isConnectedNetwork(this)) {
            Intent intent = getIntent();

            if (intent.getAction().equals(SHOW_MESSAGES_ACTION)) {
                //TODO: show message fragment

            } else {
                showThemesFragment();
            }
        }

        addOnConnectionListener(this);
        startListenInternetState();
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.themes_menu, menu);

        MenuItem theme = menu.findItem(R.id.action_black_theme);
        theme.setChecked(PreferencesHelper.getTheme(getApplicationContext()) == Constants.THEME_BLACK);

        return super.onCreateOptionsMenu(menu);

    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        //Изменяем основную тему приложения
        if (item.getItemId() == R.id.action_black_theme) {

            boolean isChecked = !item.isChecked();
            item.setChecked(isChecked);

            if (isChecked) {
                PreferencesHelper.setTheme(getApplicationContext(), Constants.THEME_BLACK);
            } else {
                PreferencesHelper.setTheme(getApplicationContext(), Constants.THEME_WHITE);
            }
            updateTheme();
            recreateActivity();

        }

        return super.onOptionsItemSelected(item);
    }

    private void setDefaultTitle() {
        setTitle(getString(R.string.app_name));
    }

    //Показываем список топиков
    private void showThemesFragment() {
        themesfragment = new ThemesListingFragment();

        getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.themes_listing_container, themesfragment)
                .commit();
    }

    //Показываем сообщения топика, добавляя новый фрагмент в бэкстек
    @Override
    public void onThemeClicked(Theme theme) {
        infoRepository.addLastViewedTheme(theme);
        Intent intent = new Intent(this, MessagesActivity.class);
        intent.putExtra(Constants.THEME_MODEL, theme);
        startActivity(intent);
    }


    @Override
    public void internetConnectionChanged(boolean connected) {
        isConnected = connected;

        if (connected) {
            if (snackbar != null) {
                snackbar.dismiss();

                //Если не было сети и не был добавлен фрагмент - добавляем
                if (themesfragment == null) {
                    showThemesFragment();
                }
            }

        } else {
            snackbar = Snackbar.make(mainContainer, R.string.connection_failed, Snackbar.LENGTH_INDEFINITE);
            snackbar.show();
        }
    }

}

File: app/src/main/java/com/example/sasham/testproject/themes/ThemesActivity.kt
package com.example.sasham.testproject.themes

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import com.example.sasham.testproject.Constants
import com.example.sasham.testproject.R
import com.example.sasham.testproject.account.AuthFragment
import com.example.sasham.testproject.base.BaseDaggerActivity
import com.example.sasham.testproject.model.User
import com.example.sasham.testproject.users.AccountFragment
import com.example.sasham.testproject.users.UsersFragment
import com.example.sasham.testproject.util.PreferencesHelper
import com.google.android.material.snackbar.Snackbar
import kotlinx.android.synthetic.main.activity_themes.*


class ThemesActivity : BaseDaggerActivity() {

    private var snackbar: Snackbar? = null
    private var isConnected = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_themes)
        setSupportActionBar(toolbar)
        setDefaultTitle()
        startListenInternetState()
        showThemesFragment()

        bottomNav.setOnNavigationItemSelectedListener {

            setDefaultTitle()
            when (it.itemId) {
                R.id.menu_themes -> {
                    showThemesFragment()
                }
                R.id.menu_users -> {
                    showUsersFragment()
                }
                R.id.menu_account -> {
                    showAccountFragment()
                }

            }
            true
        }
    }

    private fun showAccountFragment() {
        val f = supportFragmentManager.findFragmentByTag(AuthFragment::class.java.simpleName)
        if (f == null) {
            supportFragmentManager
                    .beginTransaction()
                    .replace(R.id.fragmentC, AuthFragment(), AuthFragment::class.java.simpleName)
                    .commit()
        } else {
            supportFragmentManager
                    .beginTransaction()
                    .replace(R.id.fragmentC, f, AuthFragment::class.java.simpleName)
                    .commit()
        }
    }


    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.themes_menu, menu)

        val theme = menu.findItem(R.id.action_black_theme)
        theme.isChecked = PreferencesHelper.getTheme(applicationContext) == Constants.THEME_BLACK

        return super.onCreateOptionsMenu(menu)

    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {

        //Изменяем основную тему приложения
        if (item.itemId == R.id.action_black_theme) {

            val isChecked = !item.isChecked
            item.isChecked = isChecked

            if (isChecked) {
                PreferencesHelper.setTheme(applicationContext, Constants.THEME_BLACK)
            } else {
                PreferencesHelper.setTheme(applicationContext, Constants.THEME_WHITE)
            }
            updateTheme()
            recreateActivity()

        }

        return super.onOptionsItemSelected(item)
    }

    private fun setDefaultTitle() {
        setToolbarTitle(R.string.app_name)
    }

    //Показываем список топиков
    private fun showThemesFragment() {

        val f = supportFragmentManager.findFragmentByTag(ThemesFragment::class.java.simpleName)
        if (f == null) {
            supportFragmentManager
                    .beginTransaction()
                    .replace(R.id.fragmentC, ThemesFragment(), ThemesFragment::class.java.simpleName)
                    .commit()
        } else {
            supportFragmentManager
                    .beginTransaction()
                    .replace(R.id.fragmentC, f, ThemesFragment::class.java.simpleName)
                    .commit()
        }
    }

    private fun showUsersFragment() {

        val f = supportFragmentManager.findFragmentByTag(UsersFragment::class.java.simpleName)
        if (f == null) {
            supportFragmentManager
                    .beginTransaction()
                    .replace(R.id.fragmentC, UsersFragment(), UsersFragment::class.java.simpleName)
                    .commit()
        } else {
            supportFragmentManager
                    .beginTransaction()
                    .replace(R.id.fragmentC, f, UsersFragment::class.java.simpleName)
                    .commit()
        }
    }

    override fun showLoading() {
        mainProgress.visibility = View.VISIBLE
    }

    override fun hideLoading() {
        mainProgress.visibility = View.GONE
    }

    fun showAccountScreen(it: User) {
        val f = AccountFragment.newInstance(it)
        supportFragmentManager
                .beginTransaction()
                .replace(R.id.fragmentC, f, UsersFragment::class.java.simpleName)
                .commit()
    }
}
